update to gym v0.26 Render API

tlpss opened the pull request:
Updates the `vecenv` classes to the new gym render API (no mode argument in `render` function, render mode is specified during init). 
To keep the `VecEnv`API backwards compatible, the mode argument is still accepted but it has to be the same as the environment `render_mode` attribute.

tlpss committed:
Merge pull request #1 from carlosluis/fix_tests

Fix tests

tlpss committed:
format

tlpss committed:
Merge pull request #2 from carlosluis/fix_tests

Fix tests

tlpss committed:
update render method for gym 0.26 API

backwards compatible (mode argument is allowed) while using the gym 0.26 API (render mode is determined at environment creation)

tlpss committed:
update tests and docs to new gym render API

araffin commented on stable_baselines3/common/env_checker.py:
```diff
@@ -285,34 +285,22 @@ def _check_spaces(env: gym.Env) -> None:
+    # render_modes = env.metadata.get("render.modes")
+    # TODO: if we want to check all render modes,
+    # we need to initialize new environments so the class should be passed as argument.
+    if env.render_mode:
```
why did you remove the available render mode declaration?
they are not longer declared in gym 0.26?

tlpss replied:
They are still there ( I think), but unless we can instantiate new class instances we cannot test any other mode than the one from the current instance.

araffin replied:
> ny other mode than the one from the current instance.

yes, I'm more talking about what we use to test with the warning (check if any render mode is declared and output a warning otherwise)

tlpss replied:
ah yes, my bad. I was too hasty apparently, will bring that check back

araffin commented on stable_baselines3/common/vec_env/base_vec_env.py:
```diff
@@ -55,10 +54,17 @@ class VecEnv(ABC):
         self.num_envs = num_envs
         self.observation_space = observation_space
         self.action_space = action_space
+        self.render_mode = render_mode
```
as quickly discussed in https://github.com/DLR-RM/stable-baselines3/pull/780#issuecomment-1302090826, we could do `self.render_mode = render_mode or "rbg_array"`, no?
and keep the rest of the code?

araffin replied:
actually, this should be done in `make_vec_env` and when creating the env from a string

tlpss replied:
> actually, this should be done in `make_vec_env` and when creating the env from a string


Yes we can do that when creating from string, but once a gym class instance is passed we can only 'take note' of the render mode I think?

araffin replied:
>  we can only 'take note' of the render mode I think?

yes and we do the assert only when render is called

araffin commented on stable_baselines3/common/vec_env/dummy_vec_env.py:
```diff
@@ -78,10 +78,10 @@ def close(self) -> None:
 
-    def get_images(self) -> Sequence[np.ndarray]:
-        return [env.render(mode="rgb_array") for env in self.envs]
+    def get_render_output(self) -> Sequence[Optional[np.ndarray]]:
```
I think we should keep `get_images` but raise an error if `render_mode != "rgb_array"`, until gym allows multiple render modes again, see https://github.com/Farama-Foundation/Gymnasium/issues/100

tlpss replied:
Works for me!

tlpss replied:
@araffin , how do you want the render method of the VecEnv to behave then? Should it raise an Error if the render_mode is not rgb_array (in the future if not one of the modes is rgb_array)? Or do you want to the Error in the get_images method?

tlpss replied:
@araffin, gentle ping

araffin replied:
>  Should it raise an Error if the render_mode is not rgb_array (in the future if not one of the modes is rgb_array)? 

probably a warning would be better? (so a warning and do nothing)
crashing someone's code because one tries to render doesn't look like a user-friendly behavior, no?

tlpss replied:
> I think we should keep get_images but raise an error if render_mode != "rgb_array"

Yes well I was trying to infer what you ment by this. 

I added the warning and reverted the name change. Let me know if there is anything else you would like to see different

araffin commented on stable_baselines3/common/vec_env/subproc_vec_env.py:
```diff
@@ -155,13 +160,12 @@ def close(self) -> None:
         self.closed = True
 
-    def get_images(self) -> Sequence[np.ndarray]:
+    def get_render_output(self) -> Sequence[Optional[np.ndarray]]:
```
same, we should keep `get_images` but raise an error if `render_mode` is not `"rgb_array"`

tlpss committed:
undo removal of render modes metatadata check

tlpss committed:
set rgb_array as default render mode for gym.make

araffin commented on stable_baselines3/common/env_util.py:
```diff
@@ -74,14 +74,18 @@ def make_vec_env(
         def _init():
             if isinstance(env_id, str):
-                env = gym.make(env_id, **env_kwargs)
+                # if the render mode was not specified, we set it to `rgb_array` as default.
```
I need to check, but is `render_mode` argument part of the gym api?

If not, then we should do that unless we know this argument exists.

tlpss replied:
@araffin you're right, it is not required for each class to have this argument in the init function (it is a guaranteed to exist though, see https://github.com/Farama-Foundation/Gymnasium/blob/f037fdea5a0adca5573aa5509bdc77bf48603416/gymnasium/core.py#L51) 
But their migration says this is the way to set the render mode. 

We could wrap it with a try except block to accomodate for environments that don't allow to specify the render mode?

araffin replied:
> it is a guaranteed to exist though

then should be fine (we might revert it back later in case it causes too much troubles)

tlpss committed:
undo changes & raise warning if not 'rgb_array'

araffin committed:
Merge branch 'fix_tests' into fix_tests

araffin commented on stable_baselines3/common/vec_env/base_vec_env.py:
```diff
@@ -162,35 +168,44 @@ def step(self, actions: np.ndarray) -> VecEnvStepReturn:
+    def get_images(self) -> Sequence[Optional[np.ndarray]]:
         """
-        Return RGB images from each environment
+        Return Render output from each environment
```
```suggestion
        Return RGB images from each environment when available
```

araffin commented on stable_baselines3/common/vec_env/base_vec_env.py:
```diff
@@ -162,35 +168,44 @@ def step(self, actions: np.ndarray) -> VecEnvStepReturn:
-        if mode == "human":
-            import cv2  # pytype:disable=import-error
+        if mode and self.render_mode != mode:
+            raise ValueError(
```
should be a warning I think (and return right after).
That way it does not crash.

araffin replied:
btw, I think we should support `mode="human"` when `render_mode="rgb_array"` because it is how it works with subprocess env anyway.

araffin commented on stable_baselines3/common/vec_env/base_vec_env.py:
```diff
@@ -162,35 +168,44 @@ def step(self, actions: np.ndarray) -> VecEnvStepReturn:
+
         else:
-            raise NotImplementedError(f"Render mode {mode} is not supported by VecEnvs")
+            # other render methods are simply ignored.
```
actually no, we should display the big image I think (if available, i.e. render_mode=`rgb_array`).
I know this might seems contradictory but this has to do with the fact that we are using multiple envs, not only one (and the limitation comes from https://github.com/Farama-Foundation/Gymnasium/issues/100)

tlpss replied:
So if I have understood you correctly, we  should return the tiled image for both 'human' and 'rgb_array'? If not let me know and I can change that again.

araffin replied:
I did the change ;) (so show the image when possible)

araffin commented on stable_baselines3/common/vec_env/base_vec_env.py:
```diff
@@ -162,35 +168,44 @@ def step(self, actions: np.ndarray) -> VecEnvStepReturn:
-            cv2.imshow("vecenv", bigimg[:, :, ::-1])
-            cv2.waitKey(1)
-        elif mode == "rgb_array":
+        mode = self.render_mode
```
hmm, should be `self.get_attr("render_mode")[0]` I guess.
you can either remove `self.render_mode` or better set it to `self.get_attr("render_mode")[0]` when initializing the envs (and so remove it anyway from the VecEnv constructor).

Currently, having `render_mode`  in VecEnv constructor, is kind of a duplicate, no?

araffin commented on stable_baselines3/common/vec_env/subproc_vec_env.py:
```diff
@@ -45,13 +45,15 @@ def _worker(  # noqa: C901
                 break
             elif cmd == "get_spaces":
                 remote.send((env.observation_space, env.action_space))
+            elif cmd == "get_render_mode":
```
you can probably re-use `cmd == "get_attr"`, no?

araffin commented on tests/test_vec_check_nan.py:
```diff
@@ -30,7 +30,7 @@ def step(action):
         return [0.0], {}
 
-    def render(self, mode="human", close=False):
+    def render(self, close=False):
```
we can even remove `close=False` I think

araffin commented on stable_baselines3/common/vec_env/dummy_vec_env.py:
```diff
@@ -89,10 +89,14 @@ def close(self) -> None:
-    def get_images(self) -> Sequence[np.ndarray]:
-        return [env.render(mode="rgb_array") for env in self.envs]
+    def get_images(self) -> Sequence[Optional[np.ndarray]]:
+        if self.envs[0].render_mode != "rgb_array":
```
why not re-use `self.render_mode` ?

araffin commented on stable_baselines3/common/vec_env/subproc_vec_env.py:
```diff
@@ -155,13 +160,12 @@ def close(self) -> None:
-            pipe.send(("render", "rgb_array"))
-        imgs = [pipe.recv() for pipe in self.remotes]
-        return imgs
+            # gather render return from subprocesses
```
I guess you can do the same check/warning as for the `DummyVecEnv` here

tlpss committed:
reuse existing render_mode attribute

tlpss committed:
return tiled images for 'human' render mode

tlpss committed:
Merge branch 'fix_tests' of https://github.com/carlosluis/stable-baselines3 into fix_tests

araffin committed:
Allow to use opencv for human render, fix typos

araffin committed:
Merge pull request #4 from tlpss/tlss/fix_tests

Allow to use opencv for human render, fix typos

araffin commented:
I think we are almost good to go, what is missing is to test and debug with different envs/situation.

There is one case that I'm not will work, when `env.render_mode == "human"`, it might crash anyway when using subprocesses, but it might work too. (need to be checked)

EDIT: the documentation need to be updated too to explain the behaviors, and recommend `rgb_array` as a good default ;)

tlpss commented:
> I think we are almost good to go, what is missing is to test and debug with different envs/situation.

could you briefly explain what you expect from this? The render API of the vecenv has not changed I think? so do you want to add new tests or should I modify existing tests?

araffin commented:
Hello,
i mainly meant checking manually with different envs and scenarii, we cannot really check that on the ci server.

tlpss commented:
> Hello, i mainly meant checking manually with different envs and scenarii, we cannot really check that on the ci server.

Okay, I can test for a few enviroments manually to see if the output is what I expect

Could you also point me to the place(s) where I need to change the docs?

araffin commented:
> Could you also point me to the place(s) where I need to change the docs?

I'm not sure yet. Probably the vec env page to explain difference in behavior vs gym api, and then the example page where we do some recording/render I think.

araffin committed:
Merge branch 'fix_tests' into tlpss/fix_tests

araffin committed:
Bug fixes and handle more cases

araffin committed:
Throw proper warnings

araffin committed:
Update test

araffin committed:
Fix new metadata name

araffin commented:
So, I gave it a try and fixed some bugs along the way.
I also found one bug in gym due to shared memory, those two snippets creates only one window (can be solved using subprocess):

cc @pseudo-rnd-thoughts

```python
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.vec_env import SubprocVecEnv


if __name__ == "__main__":
    n_envs = 2
    env = make_vec_env(
        "CartPole-v1",
        n_envs,
        # vec_env_cls=SubprocVecEnv,
        env_kwargs=dict(render_mode="human")
    )

    env.reset()
    env.render()

    while True:
        env.step([env.action_space.sample() for _ in range(n_envs)])
        env.render()

```


```python
import gym

env1 = gym.make("CartPole-v1", render_mode="human")
env2 = gym.make("CartPole-v1", render_mode="human")

envs = [env1, env2]

for env in envs:
    env.reset()


while True:
    for env in envs:
        env.step(env.action_space.sample())
        env.render()

```

tlpss commented:
great @araffin, I'll have a look at the docs.

araffin committed:
Merge branch 'fix_tests' into tlpss/fix_tests

tlpss commented:
> great @araffin, I'll have a look at the docs.

@araffin 
I think we might want to introduce a new page in the docs to explain the differences between the VecEnv interface and the new gym v0.26 interface? 
And then we probably need to update the examples section and add an explicit example to show the difference between interacting with a gym>0.26 interface and a vecenv interface?

Do you think of other places that should be changed? 
and are there any tests for the code snippets in the docs or do we have to test them manually?

araffin committed:
Fixes in vec recorder

araffin commented:
> I think we might want to introduce a new page in the docs to explain the differences between the VecEnv interface and the new gym v0.26 interface?

why not a section in the vec env page?
it's gonna be too big?

>  and are there any tests for the code snippets in the docs or do we have to test them manually?

we need to test them manually (too long to run at every push, see https://github.com/DLR-RM/stable-baselines3/issues/14)

araffin committed:
Merge branch 'fix_tests' into tlpss/fix_tests

araffin committed:
Merge branch 'fix_tests' into tlpss/fix_tests

araffin committed:
Merge branch 'fix_tests' into tlpss/fix_tests

araffin commented:
I'm not sure Github notification worked as I edited the original post.
@pseudo-rnd-thoughts : https://github.com/carlosluis/stable-baselines3/pull/3#issuecomment-1356863808

if you have already seen it, then please ignore my message.

araffin committed:
Merge branch 'fix_tests' into tlss/fix_tests

araffin committed:
Merge branch 'fix_tests' into tlpss/fix_tests

araffin committed:
Merge branch 'fix_tests' of github.com:tlpss/stable-baselines3 into tlpss/fix_tests

araffin committed:
Add doc of VecEnv vs Gym API

araffin committed:
Add render test

araffin commented:
@tlpss I added the doc and will merge this branch (I tested it in different scenarios and also added tests that are run only when display is available).
In the meantime, could you read what I wrote and check that nothing is missing/things are clear?
(you can do a PR or use this one if you have suggestions/remarks)

tlpss commented:
> @tlpss I added the doc and will merge this branch (I tested it in different scenarios and also added tests that are run only when display is available)

great!

> could you read what I wrote and check that nothing is missing/things are clear?

I think all differences between the gym API and VecEnv API are clearly listed. 

Only suggestion : I would  add a reference to this section in the quickstart section of the docs?

araffin commented:
> Only suggestion : I would add a reference to this section in the quickstart section of the docs?

Good point, will do that =)

pseudo-rnd-thoughts commented:
> So, I gave it a try and fixed some bugs along the way. I also found one bug in gym due to shared memory, those two snippets creates only one window (can be solved using subprocess):
> 
> cc @pseudo-rnd-thoughts
> 
> ```python
> from stable_baselines3.common.env_util import make_vec_env
> from stable_baselines3.common.vec_env import SubprocVecEnv
> 
> 
> if __name__ == "__main__":
>     n_envs = 2
>     env = make_vec_env(
>         "CartPole-v1",
>         n_envs,
>         # vec_env_cls=SubprocVecEnv,
>         env_kwargs=dict(render_mode="human")
>     )
> 
>     env.reset()
>     env.render()
> 
>     while True:
>         env.step([env.action_space.sample() for _ in range(n_envs)])
>         env.render()
> ```
> 
> ```python
> import gym
> 
> env1 = gym.make("CartPole-v1", render_mode="human")
> env2 = gym.make("CartPole-v1", render_mode="human")
> 
> envs = [env1, env2]
> 
> for env in envs:
>     env.reset()
> 
> 
> while True:
>     for env in envs:
>         env.step(env.action_space.sample())
>         env.render()
> ```

Sorry, I must have missed this. 

I will create an issue for it, I wonder if it works with vectorised gym environments
